File: baekjoon/shim/Java/단계별로_풀어보기/스택/n_9012.java
package Study_2022.baekjoon.shim.Java.단계별로_풀어보기.스택;

import java.io.*;
import java.util.*;

public class n_9012 {
    public static void main(String[] args)throws IOException{
        BufferedReader br=new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw=new BufferedWriter(new OutputStreamWriter(System.out));
        int n=Integer.parseInt(br.readLine());
        for(int i=0; i<n; i++){
            Stack<Character>stack=new Stack<>();
            boolean b=false;
            String st=br.readLine();
            for(int j=0; j<st.length(); j++){
                if(st.charAt(j)=='(')
                    stack.add('(');
                else{
                    if(stack.empty()){
                        b=true;
                        break;
                    }
                    else
                        stack.pop();
                }
            }
            if(b||!stack.empty())
                bw.write("NO\n");
            else
                bw.write("YES\n");
        }
        bw.flush();bw.close();
    }
}

File: baekjoon/shim/Python/단계별로_풀어보기/스택/n_9012.py
import sys
input=sys.stdin.readline
n=int(input())
for _ in range(n):
    l=list(input().strip())
    s=[]
    b=False
    for i in l:
        if i=='(':
            s.append('(')
        else:
            if len(s)==0:
                b=True
                break
            else:
                s.pop()
    if b or len(s)!=0:
        print("NO")
    else:
        print("YES")